### ISSUE TITLE
BIL Lifting division with negative integers

### ISSUE THREAD

**Issue by icemonster** [BinaryAnalysisPlatform/bap#961]
Lifting a simple program like
```
int main(int argc, char const *argv[])
{
	int c = 10;
	int m = -3;
	int res = c / m;
	printf("%d\n", res);
	return 0;
}
```
Seems to yield incorrect BIL instructions for the division part. Specifically, you do this check right after computing a signed division:
``` If(EQ(HIGH(0x20, Var("divXXX", Imm(0x40))), Int(0x0, 0x20)), (), CpuExn(0x0)) ``` which is, I think, incorrect for negative numbers. This seems to be checking if the signed result (quotient) is too large for the destination but we always use numbers in two complement right? Values like Var("divXXX", Imm(0x40)) are represented as bit-vectors of size 64bits, so 0xFFFFFFFFFFFFFFFF will fail that check even though its a valid number (-1). 

Am I missing anything here? Does HIGH consider the high bits of Var to be zero if Var is -1?

**Comment by ivg** [comment]
Let's simplify the program even more and consider only the division instruction, which is encoded as `f7 7d f8` in your bigger example,  here is the full bil output of this 

```fortran
$ bap-mc 'f7 7d f8' --show-insn=asm --show-bil
idivl -0x8(%rbp)
{
  if ((if high:1[mem[RBP - 8, el]:u32]
         then pad:64[mem[RBP - 8, el]:u32] | 0xFFFFFFFF00000000
         else pad:64[mem[RBP - 8, el]:u32]) = 0) {
    cpuexn (0)
  }
  #1 := (low:32[RDX].low:32[RAX]) /$ (if high:1[mem[RBP - 8, el]:u32]
                                        then pad:64[mem[RBP - 8, el]:u32] | 0xFFFFFFFF00000000
                                        else pad:64[mem[RBP - 8, el]:u32])
  #2 := (low:32[RDX].low:32[RAX]) %$ (if high:1[mem[RBP - 8, el]:u32]
                                        then pad:64[mem[RBP - 8, el]:u32] | 0xFFFFFFFF00000000
                                        else pad:64[mem[RBP - 8, el]:u32])
  if (high:32[#1] = 0) {
    
  }
  else {
    cpuexn (0)
  }
  #3 := low:32[#2].low:32[#1]
  RAX := pad:64[extract:31:0[#3]]
  RDX := pad:64[extract:63:32[#3]]
  CF := unknown[bits]:u1
  OF := unknown[bits]:u1
  SF := unknown[bits]:u1
  ZF := unknown[bits]:u1
  AF := unknown[bits]:u1
  PF := unknown[bits]:u1
}
```

as you can see there is also a check right _before_ the division, which correctly handles negative numbers.

**Comment by ivg** [comment]
On the other hand, `bap ./example --run` confirms your version, we indeed are triggering a division by zero with a negative value. 


I'll look into it, thanks.

**Comment by icemonster** [comment]
Thank you for your really fast answer.

The output for the ```bap-mc``` command that you posted seems to be from a newer version of bap. I'm using the latest stable release (1.6.0) where ```bap-mc``` shows the same fault as bap --run.
I just noticed the pull request mentioning a BAP 2.0 release coming out next October (with very exciting stuff, I must say). Is that the one you used when you ran that command? Not sure if I should switch to 2.0 before it becomes stable.

**Comment by ivg** [comment]
> Is that the one you used when you ran that command?

Yep, I was showing the output of BAP 2.x but I don't think it differs much from 1.7 modulo the temporary variables. The code the generates this BIL was written back in 2002, the buggy code was also written at that time, but was explicitly commented out with a note that it is incorrect :D At some point of time it was uncommented... morale, don't write wrong code, even in comments :)

We're now trying to understand, how the bug survived throughout all our tests. But this is a completely different story.

The fix is submitted in #962, so that it will land before 2.0 (and will be released with 1.7). Speaking of 2.0, the 1.x interface is subsumed by 2.0 so all code that uses 1.7 interface will still work in BAP 2.0. Well at least it will compile. So you can safely switch.

**Comment by icemonster** [comment]
Ahah I'm glad you found the origin of the problem!

I'm developing an auto patching system that uses symbolic execution and I'm using BIL as my intermediate language.  After a long time debugging that simple program I finally realized it wasn't my code's fault xD I suppose multiplication with negative integers is not very common because I've been testing a few vulns from the DARPA CGC dataset for the past few weeks and they seem to run fine.

Thank you so much for your time!
